### PR TITLE
feat: send list classes

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implements a new `Send_List` class for abstacting ESP-specific list and sublist data into a single schema. Also implements a `Send_Lists` class for shared util methods and REST API routes (the latter currently won't work because this PR doesn't include the `get_send_lists()` methods on each provider class).

To support the new REST API routes, moves the `newsletters/v1` namespace and the API permissions callback to the main `Newspack_Newsletters` class, so these can be shared outside of the `Newspack_Newsletters_Subscription` class.

### How to test the changes in this Pull Request:

1. Confirm that all tests pass (and review the tests for expected `Send_List` behavior)
2. In `wp_shell`, try the new constructor:

```bash
> $send_list = new Newspack\Newsletters\Send_List( [ 'id' => '123', 'name' => 'My Send List', 'type' => 'list', 'entity_type' => 'something', 'count' => 100, 'provider' => 'mailchimp' ] )
= Newspack\Newsletters\Send_List {#14997
    +"provider": "mailchimp",
    +"type": "list",
    +"entity_type": "something",
    +"id": "123",
    +"name": "My Send List",
    +"count": 100,
    +"value": "123",
    +"label": "[SOMETHING] My Send List (100 contacts)",
  }
```

3. Try out the new methods:

```bash
> $send_list->get_id()
= "123"

> $send_list->get_label()
= "[SOMETHING] My Send List (100 contacts)"

> $send_list->set_name( 'My Send List (renamed)' )
= "My Send List (renamed)"

> $send_list->get_name()
= "My Send List (renamed)"

> Newspack\Newsletters\Send_Lists::matches_search( 'renamed', [ $send_list->get( 'name' ) ] )
= true

> Newspack\Newsletters\Send_Lists::matches_id( [ 123 ], $send_list->get( 'id' ) )
= true

> $send_list->to_array()
= [
    "provider" => "mailchimp",
    "type" => "list",
    "entity_type" => "something",
    "id" => "123",
    "name" => "My Send List",
    "count" => 100,
    "label" => "[SOMETHING] My Send List (100 contacts)",
    "value" => "123",
  ]
```

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207992056607959

### CODE REVIEW

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: It'd be nice to have the accepted values for this assoc array in the docblock. E.g.:

```
	 * @param array $config {
	 *   The configuration for the Send_list.
	 *
	 *   @type string $provider The slug of the ESP for which this list or sublist originates.
	 *   @type string $type     The type of list. Can be either 'list' or 'sublist'. If the latter, must specify a `parent` property.
```

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: Following the steps from the `Subscription_List`, I wonder if it's worth being more explicit using named methods like `get_provider()` rather than `get( 'provider' )`.

I understand it requires more boilerplate but the encapsulation might be better in the long run.

Maybe @leogermani could share his thoughts?

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: NIT suggestion: the `WP_Error` class provides a nice API with `$errors->add( $code, $message )` and `$errors->has_errors()`. I find it better than handling this with try/catch.

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: We might want to return an error here.

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: Given this is always dynamic, based on the Send_List config, is there a specific reason we're setting and storing? A `get_label_and_value()` would work well and not require every config change to call this method for an update.

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
The class looks great and I love how clear the tests are!

I have a few comments:

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: I stored them as properties mainly because these are for the `Autocomplete` component in the editor JS, and having them automatically generated and set by the constructor means that I never have to make sure they're set before using them there. But you're right that if we ever update any of the properties that are used to generate these, we would need to regenerate the dynamic values. Let me think about this a bit more.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: Great idea! Moved this info to the docblock in 7f4906e.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: I agree. 7f4906e adds the following methods:

- `get_errors`
- `get_provider`
- `get_id`
- `get_type`
- `get_entity_type`
- `get_name`
- `get_local_name` (this isn't used anywhere currently, but I included local name so we can have the option to inherit the local name from the `Subscription_List` created from the same entity)
- `get_label`
- `get_value`
- `set_name`
- `set_local_name`
- `get_parent_id` (3469cf4)
- `get_count` (3469cf4)
- `set_count` (f103484)

This makes it clearer which properties you might need to get and which can be set after the object has been constructed. e.g. you can't reset `id` since it's a required property. I kept the `get` and `set` methods around as private helpers which also helps with standardizing the error handling.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: Good idea. 7f4906e refactors the error handling to use a `WP_Error` object for this instead of an array of strings. The constructor will no longer throw an exception, but will add error data that can be fetched all at once using the new `get_errors()` method. `get_*` and `set_*` methods will also return any applicable errors in addition to adding their own messages to the class instance's errors object.

Comment (dkoo) [reply includes/class-send-list.php]: #1631 will need to be refactored to take advantage of these improvements, but I'll hold off on that until this PR is merged.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: Done in 7f4906e!

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: 7f4906e introduces `get_label()` and `get_value()` methods which will either retrieve the stored values (if passed in the constructor config) or dynamically generate them. 7f4906e also takes inspiration from @leogermani's `Subscription_List` object with a `to_array()` method which will return the send list's data as an associative array. This gives us a chance to check whether we need to generate the dynamic values and append them to the response, and it also will let us tailor the data we pass back as REST responses (to be handled in #1631).

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: This leaves the impression that a get/set failing has the entire Send_List object in a broken state, which I don't think should be interpreted that way. Non-blocking, but I think it's fine to just return the error to the caller without modifying the object error.

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php]: Won't these conditionals defeat the purpose of updating these values with the latest config data?

File: includes/class-send-list.php
Comment (miguelpeixe) [inline includes/class-send-list.php:299]: Same as in `to_array()`, why bail if the value exists if we want it up to date with dynamic data?

File: includes/class-send-list.php
Comment (miguelpeixe) [reply includes/class-send-list.php]: I failed to recognize that this is the constructor 🤦 

I think we can still leverage the WP_Error class to manage multiple errors but it's best practice to maintain the throw here... Tough decision because I really like how you did it 😅

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: My thought with allowing the `label` and `value` properties to be set in the constructor was that we might want to manually set those values to override the dynamically generated ones. But I'm not actually doing that anywhere in the code, so maybe that's overkill?

The way it works now, if you don't create the `Send_List` with `label` or `value` props on the config object, then they'll always be dynamically generated when you call `to_array()`. Note how `get_label()` and `get_value()` don't set the generated values, they just return them.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: Ah, no problem! 7a3d8b3 restores the throw but keeps the use of `WP_Error` to manage the errors.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: If this isn't needed then we can remove `label` and `value` from the schema and remove all the conditionals. Otherwise we can keep as-is to retain the future option for manual overrides.

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: 7a3d8b3 simplifies this by just returning the error.

File: includes/class-send-list.php
Comment (miguelpeixe) [reply includes/class-send-list.php]: Aaaah I see now! It makes sense to me, I think it's a nice aspect, not overkill.

Comment (dkoo) [reply includes/class-send-list.php]: Awesome, I wasn't sure myself! Let's keep it, then 😄

**Review by miguelpeixe — APPROVED** [review]
Thank you for the revisions! Class and tests look great!

Let's get this in to review it in action 😎

File: includes/class-send-list.php
Comment (leogermani) [inline includes/class-send-list.php]: Setting the property directly into an undeclared object property has the downside of letting this prop public.

So I can still access `$send_list->label` which would skip our getter.

We could come up with some ideas to work around that, maybe using [overloading](https://www.php.net/manual/en/language.oop5.overloading.php) somehow, but I think the easiest thing to do is to store all properties inside one `private $config` property, which will be an array holding all the properties...

makes sense?

**Comment by miguelpeixe** [comment]
@leogermani reviewed the same time as me 😄

File: tests/test-send-lists.php
Comment (leogermani) [inline tests/test-send-lists.php:33]: Nice!

When testing methods like this it is always desirable to test some reallly unhappy tests.

For example, what's the expected return if I pass NULL as one of the parameters? or an Object of any kind? or an integer, etc.

File: includes/class-send-list.php
Comment (leogermani) [inline includes/class-send-list.php:16]: Maybe too verbose.. but just trying to add some rationale for our future selves... 

```suggestion
 * Class used to represent one Send_List.
 * 
 * A Send List is any list that can be used as a segment of contacts that you can send a Newsletter campaign to
 * The difference between Send_List and Subscription_List is that Send Lists are all the existing lists, audiences, tags, groups, segments, etc in the provider. We don't store them locally. Some of these can also be Subscription Lists, but not all are.
```

**Review by leogermani — CHANGES_REQUESTED** [review]
This is looking pretty good Derrick! Thanks for all the work building it!

I left only one comment regarding how we store the properties in the object. Other than that this is awesome!

File: includes/class-send-list.php
Comment (dkoo) [reply includes/class-send-list.php]: Ah, I didn't realize that! 69e8232 refactors to use a single protected `$config` property instead of the separate undeclared properties. This also makes the `to_array()` method a bit simpler.

File: tests/test-send-lists.php
Comment (dkoo) [reply tests/test-send-lists.php:33]: Great idea! f9bd959 adds some of these tests (as well as some missing handling for those weird cases!).

Comment (leogermani) [reply tests/test-send-lists.php:33]: That's the beauty of writing tests :)

**Review by leogermani — APPROVED** [review]

**Review by leogermani — APPROVED** [review]

**Pull request merged**
